packages/moonstone/VirtualList/VirtualListBase.js

```diff
@@ -482,82 +546,129 @@ const VirtualListBaseFactory = (type) => {
+								nextIndex = this.findNearestSpottableItemInExtent(currentIndex, candidateExtent);
+								animate = (wrap === true);
+								isWrapped = true;
+							}
+						}
 
-				for (let i = currentIndex + 2; i < dataSize; i++) {
-					if (!isItemDisabled(i)) {
-						nextIndex = i;
-						break;
+						// If there is no item which could get focus forward,
+						// we need to set restriction option to `self-first`.
+						if (nextIndex === -1) {
+							this.setRestrict(false);
+						}
+					} else {
+						nextIndex = this.findNearestSpottableItemInExtent(currentIndex, spottableExtent);
 					}
 				}
-
-				// If there is no item which could get focus forward,
-				// we need to set restriction option to `self-first`.
+			} else { // isBackward
+				nextIndex = this.findSpottableItemWithPositionInExtent(currentIndex - 1, -1, currentIndex % dimensionToExtent);
 				if (nextIndex === -1) {
-					this.setRestrict(false);
-				}
-			} else if (isBackward) {
-				// See if the next item is spottable then delegate scroll to onFocus handler
-				if (currentIndex > 0 && !isItemDisabled(currentIndex - 1)) {
-					return false;
-				}
+					spottableExtent = this.findSpottableExtent(currentIndex, false);
+
+					if (spottableExtent === -1) {
+						if (wrap && !repeat) {
+							const candidateExtent = this.getExtentIndex(this.findSpottableItem(dataSize - 1, -1));
+
+							// If currentExtent is equal to candidateExtent,
+							// it means that the current extent is the only spottable extent.
+							// So we find nextIndex only when currentExtent is different from candidateExtent.
+							if (currentExtent !== candidateExtent) {
+								nextIndex = this.findNearestSpottableItemInExtent(currentIndex, candidateExtent);
+								animate = (wrap === true);
+								isWrapped = true;
+							}
+						}
 
-				for (let i = currentIndex - 2; i >= 0; i--) {
-					if (!isItemDisabled(i)) {
-						nextIndex = i;
-						break;
+						// If there is no item which could get focus forward,
+						// we need to set restriction option to `self-first`.
+						if (nextIndex === -1) {
+							this.setRestrict(false);
+						}
+					} else {
+						nextIndex = this.findNearestSpottableItemInExtent(currentIndex, spottableExtent);
 					}
 				}
-
-				// If there is no item which could get focus backward,
-				// we need to set restriction option to `self-first`.
-				if (nextIndex === -1) {
-					this.setRestrict(false);
-				}
-			} else {
-				return false;
 			}
 
-			if (nextIndex !== -1 && (firstIndex > nextIndex || nextIndex >= firstIndex + numOfItems)) {
-				// When changing from "pointer" mode to "5way key" mode,
-				// a pointer is hidden and a last focused item get focused after 30ms.
-				// To make sure the item to be blurred after that, we used 50ms.
-				setTimeout(() => {
-					target.blur();
-				}, 50);
+			if (nextIndex !== -1) {
+				if (firstIndex <= nextIndex && nextIndex < firstIndex + numOfItems) {
```

Comment:
Shouldn't this be visible indices? please test the case.

Reply:
It works okay. `jumpToSpottableItem` returns `false` in this case.